Feature/204473/Create-Tpr-Header-Search-Component

Max-Emery opened the pull request:
In this PR:
-Created TPR header search component, a sub component of TPR header bar. 
-Implemented the accessible-autocomplete component to support result generation into dropdown box
-Added examples of header search bar to 'TPRHeaderLockUp' partial view and TPR header and footer page

max-tpr committed:
Html generation of tag complete

max-tpr committed:
Compontent renamed to match user story
#204473

max-tpr committed:
More names updated, some basic styling added

max-tpr committed:
Styling for header search improved, generation of html elements moved to header componenent due to being a sub component

max-tpr committed:
Svg image added to search component
#204473

max-tpr committed:
accessible drop down box and basic styling implemented

max-tpr committed:
Styling matching to existing search bar

max-tpr committed:
Removal of unused html generation, search bar display option added, styling adjustments

max-tpr committed:
Naming convention changed to reflect tpr component, style tag removed

max-tpr committed:
update target element name

max-tpr committed:
Responsivness improved, javascript search functionality added with static json to search. Action path of form submission no longer hard coded.

max-tpr committed:
Example search bars added to header and footer page.

max-tpr committed:
Removed commented out code

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/uSync/v9/ContentTypes/headersearch.config:
This file is from where you changed your mind and it doesn't need to be committed. In the Umbraco backoffice go to "Settings > uSync > Everything > Export > Clean export" to remove it.

Max-Emery replied:
Amended

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/wwwroot/js/autocomplete.js:
```diff
@@ -0,0 +1,28 @@
+        accessibleAutocomplete({
+
+            element: searchBar,
+            id: 'easy-autocomplete',
```
I think an id like `tpr-header-search-autocomplete` would be more appropriate.

Max-Emery replied:
Updated as suggested

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/wwwroot/js/autocomplete.js:
```diff
@@ -0,0 +1,28 @@
+            element: searchBar,
+            id: 'easy-autocomplete',
+            source: function (query, populateResults) {
+                fetch('/data.json')
```
Hard-coding `fetch('/data.json')` means this file is only ever useful in the example app, but I think it could be useful to include it in the package along with the header search box component. For that to work your tag helper would need to support an attribute like `autocomplete-url` which is rendered in the HTML as `data-autocomplete-url`. This code could then look for that attribute on `searchBar` to get the URL.

Max-Emery replied:
Amended

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/wwwroot/js/autocomplete.js:
```diff
@@ -0,0 +1,28 @@
+                        populateResults(results.slice(0,5))                 
+                    })
+                    .catch(error => {
+                        console.error('An error has occured with your fetch operation:', error);
```
It would be useful to say it's the fetch in the `tpr-header-search` component.

Max-Emery replied:
Updated

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/wwwroot/js/autocomplete.js:
```diff
@@ -0,0 +1,28 @@
+                    });
+        },
+            minLength: 2,
+            placeholder: 'Search'
```
This needs to be localisable (ie translatable into Welsh). You should be able to pick this value up from the `placeholder` attribute on the `input[type=search]` element.

Max-Emery replied:
Amended

sussexrick commented on ThePensionsRegulator.Frontend/HtmlGeneration/ComponentGenerator.TprHeaderBar.cs:
```diff
@@ -83,6 +85,76 @@ public virtual TagBuilder GenerateTprHeaderBar(TprHeaderBar tprHeaderBar)
+                    autoComplete.AddCssClass("tpr-autocomplete");
+                    searchField.InnerHtml.AppendHtml(autoComplete);
+
+                    var autoCompleteContainer = new TagBuilder("div");
```
This needs to be an `<input type="search" class="govuk-input" />` element (or there needs to be one in here). Currently your component is entirely reliant upon JavaScript, which isn't allowed in the Government Service Standard. Because it starts as a `div` the search box isn't even visible with JavaScript off. Autocomplete is always going to need JavaScript, but without JavaScript I would expect to see a fully-functional search box _without_ autocomplete.

Max-Emery replied:
Updated as suggested

sussexrick commented on ThePensionsRegulator.Frontend/HtmlGeneration/ComponentGenerator.TprHeaderBar.cs:
```diff
@@ -83,6 +85,76 @@ public virtual TagBuilder GenerateTprHeaderBar(TprHeaderBar tprHeaderBar)
             }
+            else
+            {
+                if (tprHeaderBar.DisplaySearchBar)
```
It would be neater to extract all of this search code into a separate method.

Max-Emery replied:
Updated as suggested

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/Views/TPRHeader.cshtml:
```diff
@@ -31,6 +34,8 @@
     }
     <title>@Model.PageHeadingOrName()</title>
     <link rel="stylesheet" href="~/css/site.css" />
+    <script src="~/js/accessible-autocomplete.min.js"></script>
```
These files are in the example app, but your component will be published in a NuGet package and consumed by other applications, and these files won't be there. You should be able to put them under `ThePensionsRegulator.Frontend/wwwroot/tpr` (`accessible-autocomplete.min.js` would go in a `wwwroot/tpr/lib/accessible-autocomplete` folder) - see https://github.com/thepensionsregulator/govuk-frontend-aspnetcore-extensions/blob/develop/docs/contributing/include-client-side-files-in-packages.md

Max-Emery replied:
Amended

sussexrick commented on ThePensionsRegulator.Frontend/Models/TprHeaderLockupModel.cs:
```diff
@@ -26,5 +26,6 @@ public class TprHeaderLockupModel
         public virtual string? Context1 { get; set; }
         public virtual string? Context2 { get; set; }
         public virtual string? Context3 { get; set; }
+        public virtual string? ActionPath { get; set; } = "en/search-result";
```
The default value `en/search-result` is specific to our public website but you're building a component that shouldn't know which site it'll be used on, so just remove this default value.

Max-Emery replied:
Default value removed

sussexrick commented on ThePensionsRegulator.Frontend/HtmlGeneration/ComponentGenerator.TprHeaderBar.cs:
```diff
@@ -83,6 +85,76 @@ public virtual TagBuilder GenerateTprHeaderBar(TprHeaderBar tprHeaderBar)
+                    var autoCompleteContainer = new TagBuilder("div");
+                    autoCompleteContainer.AddCssClass("tpr-autocomplete-container");
+                    autoCompleteContainer.Attributes.Add("id", "eac-container-GlobalSearchInputBox");
+                    autoComplete.Attributes.Add("placeholder", "Search");
```
The placeholder text "Search" needs to be localisable (ie translatable) into Welsh. To do that, make the value come from a property on the tag helper. Then it's up to the consuming application to handle the translation and configure the desired text.

Max-Emery replied:
Updated as suggested

sussexrick commented on ThePensionsRegulator.Frontend/HtmlGeneration/ComponentGenerator.TprHeaderBar.cs:
```diff
@@ -83,6 +85,76 @@ public virtual TagBuilder GenerateTprHeaderBar(TprHeaderBar tprHeaderBar)
+                    var button = new TagBuilder("button");
+                    button.Attributes.Add("type", "submit");
+                    button.Attributes.Add("class", "searchButton");
+                    button.Attributes.Add("aria-label", "Search");
```
The ARIA label text "Search" needs to be localisable (ie translatable) into Welsh. To do that, make the value come from a property on the tag helper. Then it's up to the consuming application to handle the translation and configure the desired text.

Max-Emery replied:
Updated as suggested

sussexrick commented on ThePensionsRegulator.Frontend/HtmlGeneration/ComponentGenerator.TprHeaderBar.cs:
```diff
@@ -83,6 +85,76 @@ public virtual TagBuilder GenerateTprHeaderBar(TprHeaderBar tprHeaderBar)
+                    form.Attributes.Add("method", "get");
+
+                    var searchField = new TagBuilder("div");
+                    searchField.AddCssClass("searchFieldWithButton");
```
Your classes should use the [BEM](https://getbem.com/introduction/) naming pattern, eg `tpr-header-search`, `tpr-header-search__button` and so on. IDs should all be in the form `tpr-header-search-*`. 

This makes it clear where they apply to, and they shouldn't clash with anything else on the page, like another search in the middle of the page that uses autocomplete with a different data source.

Max-Emery replied:
Updated as suggested

sussexrick commented on ThePensionsRegulator.Frontend/Styles/_tpr-header-search.scss:
```diff
@@ -0,0 +1,226 @@
+    padding: 0.5em 0.5em;
+    text-align: center;
+    display: flex;
+    flex: 0 0, auto;
```
There shouldn't be a comma in `flex: 0 0 auto`. You're specifying `text-align` and `flex-direction` twice.

Max-Emery replied:
Amended

sussexrick replied:
You're still specifying text-align and flex-direction twice.

Max-Emery replied:
Duplicate code now removed

sussexrick commented on ThePensionsRegulator.Frontend/Styles/_tpr-header-search.scss:
```diff
@@ -0,0 +1,226 @@
+}
+
+.tpr-autocomplete {
+    width: 313.948px
```
This is very precise! What's the reason?

Max-Emery replied:
There isn't a specific reason for this value. It was the default size from the original CSS file packaged with accessible-autocomplete.

sussexrick commented on ThePensionsRegulator.Frontend/Styles/_tpr-header-search.scss:
```diff
@@ -0,0 +1,226 @@
+    vertical-align: middle;
+}
+
+.searchbox[type=text]:focus {
```
The search input has a black border that the current TPR website doesn't have.

When you tab to the search input and search button they should get a GOV.UK-style focus ring. Try tabbing to form controls on the other example pages to see what I mean.  You shouldn't need to define a custom focus style like this.

On the search input you can apply class `govuk-input` to get it. On the button I think you need the `govuk-focused-box` SASS mixin. See https://frontend.design-system.service.gov.uk/sass-api-reference/#govuk-focused-box

Also when the button is focussed it has an odd shape:

![image](https://github.com/user-attachments/assets/017f4876-165a-4923-a7b9-2eba47b12f50)

I found that if you set the `height` and `width` to `2rem` and change the `viewBox` attribute to `0 0 21 6` it makes it better.

Max-Emery replied:
Button updated

sussexrick commented on ThePensionsRegulator.Frontend/Styles/_tpr-header-search.scss:
```diff
@@ -0,0 +1,226 @@
+    outline-style: solid;
+}
+
+svg {
```
This `svg` style is going to apply to any SVG image included in the page in the same way. You need to make it specific to the SVG in the header search component, eg by using a class like `tpr-header-search__button-image`.

Max-Emery replied:
class added

sussexrick commented on ThePensionsRegulator.Frontend/Styles/_tpr-header-search.scss:
```diff
@@ -0,0 +1,226 @@
+}
+
+.autocomplete__input {
+    background-color: #fff;
```
All colours should use variables from the `_tpr-variables.scss` file. For example this one will be `$tpr-colour-white`. You may find available TPR styles are slightly different to what you have, eg a different shade of black. That's OK - go with TPR colours.

Max-Emery replied:
Updated

sussexrick commented on ThePensionsRegulator.Frontend/Styles/_tpr-header-search.scss:
```diff
@@ -0,0 +1,226 @@
+    padding: 0;
+    width: 85%;
+    width: calc(100% - 4px);
+    font-family: "Segoe UI Symbol";
```
Use `font-family: $govuk-font-family` to fit in with fonts elsewhere.

Max-Emery replied:
Updated

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/Views/TPRHeader.cshtml:
```diff
@@ -181,6 +186,26 @@
+            </div>
+            <tpr-header-bar>
+                <tpr-header-bar-logo alt="@headerLogoAlt" href="@(settings?.Value<Link>("tprHeaderLogoHref")?.Url)" />
+                <tpr-header-search action="/@languageTwoLetterCode/@actionpath"></tpr-header-search>
```
I like the examples you've added, but the search boxes do go to a 404 page. In this example app we usually send everything to `/success/`.

Max-Emery replied:
Updated

sussexrick commented on ThePensionsRegulator.Frontend/TagHelpers/TprHeaderSearchTagHelper.cs:
```diff
@@ -0,0 +1,29 @@
+
+            var content = await output.GetChildContentAsync();
+
+            headerSearchContext.SetSearch(output.Attributes.ToAttributeDictionary(), true, ActionPath);
```
`ActionPath` may be null here, so you need to change the definition of `SetSearch` to accept `string?`

Max-Emery replied:
Updated as suggested

sussexrick commented on ThePensionsRegulator.Frontend/TagHelpers/TprHeaderBarContext.cs:
```diff
@@ -19,6 +20,10 @@ internal class TprHeaderBarContext
         public AttributeDictionary? ContentAttributes => _content?.Attributes;
         public IHtmlContent? Content => _content?.Content;
         public bool ContentAllowHtml => _content?.AllowHtml ?? false;
+        public bool DisplaySearchBar => _search?.DispaySearchBar ?? false;
```
I don't think this is a "bar". It appears within the "header bar". I would call this `DisplaySearch`.

Max-Emery replied:
Amended

sussexrick replied:
You've updated `TprHeaderBarContext` but not `TprHeaderBar.DisplaySearchBar`.

Max-Emery replied:
TprHeaderBar.DisplaySearchBar also updated

sussexrick commented on ThePensionsRegulator.Frontend/Styles/_tpr-header-search.scss:
```diff
@@ -0,0 +1,226 @@
+    position: relative;
+}
+
+::placeholder {
```
This is going to affect every piece of placeholder text in every form element. This isn't the place to do something like that, so it should be specific to this `autocomplete_input::placeholder`. Ideally it wouldn't be `0.8em` which is smaller than the user's chosen font size, but see how `1rem` looks.

Max-Emery replied:
Updated as suggested

sussexrick commented on ThePensionsRegulator.Frontend/Styles/_tpr-header-search.scss:
```diff
@@ -0,0 +1,226 @@
+    vertical-align: bottom;
+}
+
+.autocomplete__wrapper {
```
All these `.autocomplete-*` styles should be in their own file because they don't only apply to the header search box. They will apply to any autocomplete used on the page.

Max-Emery replied:
Updated, moved to new file

sussexrick replied:
Your new file still has `header-search` in the name, so it's still part of the specific component. It could just be `_tpr-autocomplete.scss`.

Also I noticed you could be using `$govuk-focus-width` instead of `3px`.

Max-Emery replied:
Amended

sussexrick commented on ThePensionsRegulator.Frontend/Styles/_tpr-header-search.scss:
```diff
@@ -0,0 +1,226 @@
+    }
+}
+
+@media(max-width: 864px) {
```
I think this needs work. On the current public website the search is visible down to around 765px. Somewhere around 990px the header bar label "Making workplace pensions work" disappears to make way for the search to continue to show.

Max-Emery replied:
Amended, mirrors current site better

sussexrick commented on ThePensionsRegulator.Frontend/Views/Shared/TPR/TPRHeaderLockup.cshtml:
```diff
@@ -11,7 +11,13 @@
+    var headerContent = Model.HeaderBarContent;  
+        
 }
+<head>
```
You can't include the JavaScript files this way. There can only be one `<head>` in a page and this partial view will be used inside `<body>`. Instead you can put a separate partial view in https://github.com/thepensionsregulator/govuk-frontend-aspnetcore-extensions/tree/develop/ThePensionsRegulator.Frontend/Views/Shared/TPR and document that it needs to be included by any sites wanting to use the search with autocomplete.

Max-Emery replied:
Separate view created and documentation updated

sussexrick replied:
This view is called `TPRHeaderSearchAutocomplete` but, similar to the SCSS file, it's not unique to the header. It could actually be used by other implementations of autocomplete on the page.

Also, you could be using this on the demo page `TPRHeader.cshtml`. Instead you have the script tags repeated in there except that the paths are different. The paths in this partial view are wrong.

Max-Emery replied:
This has been amended

sussexrick replied:
This is better now, but `autocomplete.js` still has several references to header search that don't need to be there. That still makes this partial view limited to the header search when it could be generic for any use of autocomplete on the page.

1. your variables `searchBars` and `searchBar` could be something like `containers` and `container` that don't make any assumptions about what the autocomplete is used for.
2. `document.querySelectorAll('.tpr-header-search__input')` could be `searchBar.querySelector('.govuk-input')` since you don't actually want multiple results from the whole page, you only want the first input within the container you've identified. With only one result you then don't need the `forEach` that follows it.
3. On line 21 the hard-coded id `tpr-header-search-autocomplete` could be found. `searchBar.querySelector(".autocomplete__input").id` should do it.
4. In the error message on line 30 you could just replace `tpr-header-search` with `autocomplete`

Max-Emery replied:
Updated 1,2,4.

**question:** id: `tpr-header-search-autocomplete` is setting the value for the id on the element, this is needed in order to generate the search box, is it better to leave it as `tpr-autocomplete__input`, or should this value for the id come from another element and be able to be set on the tag helper?

sussexrick commented on ThePensionsRegulator.Frontend/Views/Shared/TPR/TPRHeaderLockup.cshtml:
```diff
@@ -11,7 +11,13 @@
     var headerLogoAlt = Model.LogoAlternativeText;
     if (string.IsNullOrEmpty(headerLogoAlt) && !string.IsNullOrEmpty(Model.LogoHref)) { headerLogoAlt = ComponentGenerator.HeaderLogoDefaultAlt; }
     var showPhaseBanner = Model.ShowPhaseBanner && !string.IsNullOrEmpty(phase) && phase != "Live";
+    var headerContent = Model.HeaderBarContent;  
```
This new variable is never used.

Max-Emery replied:
Removed variable

sussexrick commented on ThePensionsRegulator.Frontend/Views/Shared/TPR/TPRHeaderLockup.cshtml:
```diff
@@ -30,14 +36,17 @@
-                <tpr-header-bar-label>@Model.HeaderBarLabel</tpr-header-bar-label>
-                <tpr-header-bar-content allow-html="@Model.HeaderBarContentAllowHtml">@Html.Raw(Model.HeaderBarContent)</tpr-header-bar-content>
+                <tpr-header-bar-label>@Model.HeaderBarLabel</tpr-header-bar-label> 
+              <tpr-header-search action="@Model.ActionPath"></tpr-header-search>
```
You've _replaced_ the existing support for content with support for search. You need an if statement that supports a choice of one or the other.

Max-Emery replied:
Added back in support for content and added choice for search

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/Models/ModelsBuilder/HeaderFooterTpr.generated.cs:
In Umbraco you could look at adding fields to `Document types > TPR > TPR header` that allow a site to configure text strings like the placeholder text and aria-label text.

Max-Emery replied:
Added fields to Umbraco

sussexrick replied:
The aliases for these should start with `tprHeader` like the other properties on the same document type. That will prevent a clash with any other placeholder and aria label settings on other document types.

Also it would be useful to say here, in either the name or the description, that this is the aria label for the search _button_:

![BackOffice view of aria label property](https://github.com/user-attachments/assets/63527f14-1369-4b1f-bef2-f8aecad949b1)

Max-Emery replied:
Properties renamed

sussexrick replied:
You need to use these properties in `UmbracoTprHeaderLockupModel.cs` too.

Max-Emery replied:
Added to file

sussexrick reviewed (changes requested):
I like what you've done @Max-Emery, this is a great first effort. It works well within the context of the example app. Now you need to make some changes so that it is:
- more consistent with how other components are built (mostly in the names and styles)
- able to be packaged up for re-use outside the example app
- localisable, so that Welsh is fully supported
- not reliant on JavaScript for the basic search

You also need to add details of your component to https://github.com/thepensionsregulator/govuk-frontend-aspnetcore-extensions/blob/develop/docs/components/tpr-header-bar.md

max-tpr committed:
Naming more consistant to other components, gov and tpr varibles used in css.
Placholder and aria-label properties are now localisable.
Html generation for tpr-header-search abstracted to seperated into own mehtod for better readability

max-tpr committed:
Duplicate name attriubte removed to avoid duplication in url when searching with JavaScript enabled.
Hardcoded fetch destination and placeholder text updated to use tag helper attributes
Html generation restricted when JavaScript is enabled to avoid duplicate input elements being generated.
govuk-input styling added to component

max-tpr committed:
svg alignment adjusted

max-tpr committed:
header bar label disappears on smaller screens

max-tpr committed:
Clean export removing unused headersearcg config file

max-tpr committed:
Removed defualt value of action path, added properties for other attributes. Removal of head element in partial view, created seperate partial view to host javascript

max-tpr committed:
JS files moved to be under correct project so can be published as part of the NuGet package

max-tpr committed:
Added Html attribute to be able to set aria-label value for input and button

max-tpr committed:
Documentation added for tpr-header-search

max-tpr committed:
Conficts resolved

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/uSync/v9/Languages/en-us.config:
Not your fault - in Umbraco en-US is a zombie language that keeps coming back to life no matter how many times you kill it. 🧟🧟🧟

Please delete it from your installation and then you can delete this file.

Max-Emery replied:
Actioned

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/uSync/v9/Content/text-input.config:
```diff
@@ -735,10 +735,12 @@
+			</Value>
+		</blocks>
+		<grid>
+			<Value Culture="de">
```
German was changed to Welsh and you're reverting it. This just needs not to change.

Max-Emery replied:
I've changed this back

sussexrick replied:
This file has a stray letter 's' in it.

Max-Emery replied:
Removed

sussexrick commented on GovUk.Frontend.Umbraco.ExampleApp/Models/ModelsBuilder/HeaderSearch.generated.cs:
I suspect you need to regenerate Models Builder models because I don't see any reference to your new fields `arialabelText` and `placeholderText`.

Max-Emery replied:
Models regenerated

sussexrick replied:
I think you need to regenerate again. On downloading your branch I'm getting a build error on a ModelsBuilder file:

<img width="585" height="39" alt="image" src="https://github.com/user-attachments/assets/a8b1beda-94ba-49f8-b016-247136797056" />

Max-Emery replied:
I've re-ran the model builder

sussexrick commented on ThePensionsRegulator.Frontend/HtmlGeneration/ComponentGenerator.TprHeaderBar.cs:
```diff
@@ -83,9 +85,96 @@ public virtual TagBuilder GenerateTprHeaderBar(TprHeaderBar tprHeaderBar)
+            divTag.AddCssClass("tpr-header-search");
+
+            var form = new TagBuilder("form");
+            form.Attributes.Add("action", $"{tprHeaderBar.ActionPath}");
```
`ActionPath` is a nullable property so this could generate `action=""`. Either `ActionPath` should be `required`, or you should avoid generating an attribute if it's blank. A form will no action will send the response the current page by default.

Max-Emery replied:
Amended

sussexrick commented on ThePensionsRegulator.Frontend/HtmlGeneration/ComponentGenerator.TprHeaderBar.cs:
```diff
@@ -83,9 +85,96 @@ public virtual TagBuilder GenerateTprHeaderBar(TprHeaderBar tprHeaderBar)
+            var autoCompleteContainer = new TagBuilder("div");
+            autoCompleteContainer.AddCssClass("tpr-autocomplete-container");
+            autoCompleteContainer.Attributes.Add("autocomplete-url", tprHeaderBar.AutoCompleteUrl);
+            autoCompleteContainer.Attributes.Add("placeholder", tprHeaderBar.SearchPlaceholderText);
```
`placeholder` isn't a valid attribute on a `div`

Max-Emery replied:
Removed from div

sussexrick commented on ThePensionsRegulator.Frontend/HtmlGeneration/ComponentGenerator.TprHeaderBar.cs:
```diff
@@ -83,9 +85,96 @@ public virtual TagBuilder GenerateTprHeaderBar(TprHeaderBar tprHeaderBar)
+
+            var autoCompleteContainer = new TagBuilder("div");
+            autoCompleteContainer.AddCssClass("tpr-autocomplete-container");
+            autoCompleteContainer.Attributes.Add("autocomplete-url", tprHeaderBar.AutoCompleteUrl);
```
`AutoCompleteUrl` is a nullable property so this could generate `autocomplete-url=""`. Either `AutoCompleteUrl` should be `required`, or you should avoid generating an attribute if it's blank.

Also `autocomplete-url` is not a valid attribute in HTML. `data-autocomplete-url` would be valid.

Max-Emery replied:
Amended

sussexrick replied:
This looks good, but VS is highlighting a stray semi-colon:

<img width="761" height="77" alt="image" src="https://github.com/user-attachments/assets/27ee16c1-3f27-4697-ad3d-dd980de4dd49" />

Max-Emery replied:
Removed

sussexrick commented on ThePensionsRegulator.Frontend/HtmlGeneration/ComponentGenerator.TprHeaderBar.cs:
```diff
@@ -83,9 +85,96 @@ public virtual TagBuilder GenerateTprHeaderBar(TprHeaderBar tprHeaderBar)
+            autoComplete.InnerHtml.AppendHtml(autoCompleteContainer);
+
+            var searchInput = new TagBuilder("input");
+            searchInput.Attributes.Add("aria-label", tprHeaderBar.SearchAriaLabel);
```
This ARIA label is for the `button`, so it shouldn't be on the `input` as well.

Max-Emery replied:
Amended

sussexrick commented on ThePensionsRegulator.Frontend/HtmlGeneration/ComponentGenerator.TprHeaderBar.cs:
```diff
@@ -83,9 +85,96 @@ public virtual TagBuilder GenerateTprHeaderBar(TprHeaderBar tprHeaderBar)
+
+            var searchInput = new TagBuilder("input");
+            searchInput.Attributes.Add("aria-label", tprHeaderBar.SearchAriaLabel);
+            searchInput.Attributes.Add("placeholder", tprHeaderBar.SearchPlaceholderText);
```
`SearchPlaceholderText` is a nullable property so this could generate `placeholder=""`. Either `SearchPlaceholderText` should be `required`, or you should avoid generating an attribute if it's blank.

Max-Emery replied:
Amended

sussexrick commented on ThePensionsRegulator.Frontend/HtmlGeneration/ComponentGenerator.TprHeaderBar.cs:
```diff
@@ -83,9 +85,96 @@ public virtual TagBuilder GenerateTprHeaderBar(TprHeaderBar tprHeaderBar)
+            searchInput.Attributes.Add("aria-label", tprHeaderBar.SearchAriaLabel);
+            searchInput.Attributes.Add("placeholder", tprHeaderBar.SearchPlaceholderText);
+            searchInput.Attributes.Add("type", "search");
+            searchInput.Attributes.Add("name", "query");
```
It would be good to make `"query"` configurable, so that the destination page can choose what query string parameter it wants to handle. It's fine to have `"query"` as a default.

Max-Emery replied:
This is now configurable

sussexrick replied:
This isn't quite working. If you try your example in the 'header and footer' example page you see that the query string now includes two versions of the parameter:

<img width="375" height="27" alt="image" src="https://github.com/user-attachments/assets/d09c966b-25de-4cb5-8b1d-177a24569722" />

Also, I think the attribute on the tag helper needs to be `input-name` not `input-query` because it's the `name` attribute that's being set. That would be consistent with other tag helpers. Follow that through the code, eg `SearchInputQuery` becomes `SearchInputName`.

Max-Emery replied:
Removed JS code that was adding extra query string value, name updated

sussexrick replied:
The JS isn't working. It no longer adds the duplicate query string parameter, but it also doesn't add the search term: 

<img width="821" height="56" alt="image" src="https://github.com/user-attachments/assets/61a176f9-bea4-4674-9e1e-9bde579ee83b" />

It does work without JS.

Also, you didn't follow the name change through the code. It's starts out as `InputName` but soon becomes `InputQuery` as soon as you pass it into `TprHeaderBarContext.SetSearch()`

Max-Emery replied:
Amended

sussexrick commented on ThePensionsRegulator.Frontend/HtmlGeneration/ComponentGenerator.TprHeaderBar.cs:
```diff
@@ -83,9 +85,96 @@ public virtual TagBuilder GenerateTprHeaderBar(TprHeaderBar tprHeaderBar)
+            button.Attributes.Add("type", "submit");
+            button.AddCssClass("tpr-header-search__button");
+            button.AddCssClass("govuk-input");
+            button.Attributes.Add("aria-label", tprHeaderBar.SearchAriaLabel);
```
`SearchAriaLabel` is a nullable property so this could generate `aria-label=""`. Either `SearchAriaLabel` should be `required`, or you should avoid generating an attribute if it's blank.

Max-Emery replied:
Amended

sussexrick commented on ThePensionsRegulator.Frontend/Models/TprHeaderLockupModel.cs:
```diff
@@ -26,5 +26,10 @@ public class TprHeaderLockupModel
         public virtual string? Context1 { get; set; }
         public virtual string? Context2 { get; set; }
         public virtual string? Context3 { get; set; }
+        public virtual bool DisplaySearch { get; init; } = true;
```
Should this really default to `true`? Most TPR websites don't have this search box, so I think you should have no default value which means it defaults to `false`.

Max-Emery replied:
Removed default

sussexrick replied:
For consistency I think `DisplaySearch` needs to be renamed to `ShowSearch`. All the other toggles for elements of the header are called `Show<thing>`.

sussexrick replied:
Also I think these four properties need to start with `Search` to clarify what they apply to. Notice how that's been done for other properties relating to `SkipLink` and `Logo` for example.

<img width="385" height="123" alt="image" src="https://github.com/user-attachments/assets/19c9a710-af6a-4452-a477-9c5bdb9a14dd" />

Max-Emery replied:
Updated

sussexrick commented on docs/components/tpr-header-bar.md:
```diff
@@ -14,6 +14,7 @@ The Pensions Regulator (TPR) uses the TPR header bar as a consistent part of the
         <a class="govuk-link" href="#">A link</a>
         <a class="govuk-link" href="#">Another link</a>
     </tpr-header-bar-content>
+    <tpr-header-search>
```
This now has attributes you could be showing.

Max-Emery replied:
Attributes added to docs

sussexrick commented on docs/components/tpr-header-bar.md:
```diff
@@ -60,6 +61,21 @@ Typically used for a menu of relevant links.
+|     `action`      |`string`| Sets action path of search form submission.                           |
+|`autocomplete-url` |`string`| Sets endpoint for autocomplete.js fetch operation.                    |
+|   `placeholder`   |`string`| Supports setting custom placeholder for generated input elements.     |
+|   `aria-label`    |`string`| Programmatically setting aria-label value of input and button elements|
```
Just the button. They're separate things so they shouldn't have the same label. Testing with NVDA the placeholder text is sufficient for the input.

Max-Emery replied:
Amended

sussexrick commented on docs/components/tpr-header-bar.md:
```diff
@@ -60,6 +61,21 @@ Typically used for a menu of relevant links.
+|   `placeholder`   |`string`| Supports setting custom placeholder for generated input elements.     |
+|   `aria-label`    |`string`| Programmatically setting aria-label value of input and button elements|
+
+TPR Header Search will display when there is no header content and DisplayHeaderSearch property is set to true. On smaller screen sizes the tpr mobile menu component will take over, and the header search will display as part of the mobile menu when expanded.
```
TPR should be in caps, and you should also explain that the mobile behaviour is not yet implemented and therefore this is currently a desktop-only component.

Max-Emery replied:
Amended

sussexrick commented on docs/components/tpr-header-bar.md:
```diff
@@ -60,6 +61,21 @@ Typically used for a menu of relevant links.
+
+TPR Header Search will display when there is no header content and DisplayHeaderSearch property is set to true. On smaller screen sizes the tpr mobile menu component will take over, and the header search will display as part of the mobile menu when expanded.
+
+TPR Header Search implements the alphagov/accessible-autocomplete component, rendering an input box and drop-down box for search resuts.
```
It would be useful to link the text `alphagov/accessible-autocomplete` to the repo.

Max-Emery replied:
Amended

sussexrick commented on docs/components/tpr-header-bar.md:
```diff
@@ -60,6 +61,21 @@ Typically used for a menu of relevant links.
+
+TPR Header Search implements the alphagov/accessible-autocomplete component, rendering an input box and drop-down box for search resuts.
+
+Views which require the tpr-header-search should also include the 'TPRHeaderSearchAutocomplete' partial view, in order to use the autocomplete functionality.
```
`<tpr-header-search>` rather than `tpr-header-search`

Max-Emery replied:
Amended

max-tpr committed:
Documentation updated, DisplaySearch defualt value removed, US language removed from umbraco, German culture varient reverted back to Welsh.

max-tpr committed:
TprHeader model regenerated to include new properties, values for tag helper now come from umbraco.

max-tpr committed:
Query is now configurable through new property and html attribute. Js updated to retrieve this value for generate autocomplete input box. Null checks introduced for generated attributes to avoid generating empty values.

max-tpr committed:
Script tags removed in place for partial view.

max-tpr committed:
CSS updated, removed duplicate code and added gov variables

max-tpr committed:
Futher duplication removed

max-tpr committed:
Docs update for input query

max-tpr committed:
Null checks improved to handle empty values

max-tpr committed:
Typo fixed

max-tpr committed:
Confilct resloved

max-tpr committed:
Typo  fixed

max-tpr committed:
Merge branch 'develop' of https://github.com/thepensionsregulator/govuk-frontend-aspnetcore-extensions into feature/204473/Create-TPR-Header-Search-Component

max-tpr committed:
Merge 'develop' into feature/204473/Create-TPR-Header-Search-Component

sussexrick commented on ThePensionsRegulator.Frontend/Styles/_tpr-header-search.scss:
```diff
@@ -0,0 +1,75 @@
+        margin-left: 50px;
+    }
+
+    .searchbox {
```
I don't think .searchbox is used anywhere in your code. If not, it can be removed.

Max-Emery replied:
Amended

max-tpr committed:
DisplaySearch updated for consistancy, Model builder re-ran, removal of character from text-input.config causing uSync error, removal of stray semi-colon. query string name fixed

max-tpr committed:
Model generator re-ran, added properties to relevant Models, added ShowSearch functionality to backoffice. Updated naming convention for properties.

max-tpr committed:
Removal of accidental addition of mobile menu models

max-tpr committed:
Empty query string fixed, autocomplete.js updated to be more compatible with other components.

max-tpr committed:
Updated InputQuery to InputName

AlexBeesley reviewed (approved):
Approving this, good work @Max-Emery. Have tested this locally and Max has responded to @sussexrick suggestions with appropriate code changes.

max-tpr committed:
Merged with develop
